### PR TITLE
Add RpcBriefcaseUtility.findOrOpen and automatically refresh SAS key for v2 checkpoints

### PR DESCRIPTION
**Pull request by paulius-valiunas**

### CODE REVIEW

File: core/backend/src/IModelDb.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@
 // cspell:ignore ulas postrc pollrc CANTOPEN
 
 import {
-  assert, BeEvent, BentleyStatus, ChangeSetStatus, ClientRequestContext, DbResult, Guid, GuidString, Id64, Id64Arg, Id64Array, Id64Set, Id64String,
+  assert, AsyncMutex, BeEvent, BeEventList, BentleyStatus, ChangeSetStatus, ClientRequestContext, DbResult, Guid, GuidString, Id64, Id64Arg, Id64Array, Id64Set, Id64String,
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:12]: these are unused.

File: core/backend/src/IModelDb.ts [hunk 2]
```diff
@@ -2533,6 +2534,20 @@ export class SnapshotDb extends IModelDb {
       snapshot.close();
       throw err;
     }
+
+    if (sasTokenExpiry) {
+      const expiresIn = Date.parse(sasTokenExpiry) - Date.now();
+
+      setTimeout(async () => {
+        const unlock = await RpcConfiguration.requestContext.requestMutex.lock();
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:2542]: this doesn't seem necessary or warranted to me. Why do you need the "mutex"?

Comment (paulius-valiunas) [reply core/backend/src/IModelDb.ts:2542]: This is still work in progress and I probably need to think about it more, but I wanted to avoid any requests being processed during the reattach. I wasn't sure if the daemon would be usable without interruptions while V2CheckpointManager.attach is being executed so I decided to play it safe. But I admit I didn't think it through very well, so I'll check if this is actually needed.

Comment (kabentley) [reply core/backend/src/IModelDb.ts:2542]: yes, the deamon will continue to work successfully during the refresh unless the previous token expires. You should request the refresh long before that happens. None of this complication for the RPC layer should be necessary.

File: core/backend/src/IModelDb.ts [hunk 3]
```diff
@@ -2524,7 +2525,7 @@ export class SnapshotDb extends IModelDb {
    * @internal
    */
   public static async openCheckpointV2(checkpoint: CheckpointProps): Promise<SnapshotDb> {
-    const filePath = await V2CheckpointManager.attach(checkpoint);
+    const { filePath, sasTokenExpiry } = await V2CheckpointManager.attach(checkpoint);
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:2528]: it seems to me that the auto-reattach should happen *inside* `V2CheckpointManager.attach` so no caller needs to worry about it.

File: core/backend/src/CheckpointManager.ts [hunk 4]
```diff
@@ -158,9 +160,23 @@ export class V2CheckpointManager {
...
+        await this.attach(checkpoint);
+      }, expiresIn / 2);
+    }
+
     return BlobDaemon.getDbFileName(args);
   }
 
+  public static detach(checkpointKey: string) {
+    if (this.reattachTimeouts.hasOwnProperty(checkpointKey))
+      clearTimeout(this.reattachTimeouts[checkpointKey]);
```
Comment (wgoehrig) [inline core/backend/src/CheckpointManager.ts:177]: Should also remove from `reattachTimeouts`, right?

File: core/backend/src/CheckpointManager.ts [hunk 5]
```diff
@@ -158,9 +160,23 @@ export class V2CheckpointManager {
 
       throw new IModelError(attachResult.result, error);
     }
+    const sasTokenExpiry = new URLSearchParams(args.auth).get("se");
+    if (sasTokenExpiry) {
+      const expiresIn = Date.parse(sasTokenExpiry) - Date.now();
+
+      this.reattachTimeouts[CheckpointManager.getKey(checkpoint)] = setTimeout(async () => {
+        await this.attach(checkpoint);
```
Comment (wgoehrig) [inline core/backend/src/CheckpointManager.ts:168]: This isn't safe - there's no guarantee the `AuthorizedClientRequestContext` in `checkpoint` will still be authorized when that timer fires (actually it's probably so far in the future it almost certainly won't...)

**Review by wgoehrig — CHANGES_REQUESTED** [review]

File: core/backend/src/CheckpointManager.ts (hunk 5 above)
Comment (paulius-valiunas) [reply core/backend/src/CheckpointManager.ts:168]: good point. And the only way to get a correct token is from the frontend? In that case, I guess force-closing the SnapshotDb is the only solution...

Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:168]: bill, isn't that why we decided that we're going to have to use client credentials to get the sas token?

Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:168]: @wgoehrig what's the plan for obtaining a token to ask for the SAS key?

File: core/backend/src/IModelDb.ts [hunk 6]
```diff
@@ -833,6 +836,17 @@ export abstract class IModelDb extends IModel {
     return undefined;
   }
 
+  public static async findOrOpen(requestContext: AuthorizedClientRequestContext, iModel: IModelRpcProps, syncMode: SyncMode): Promise<IModelDb> {
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:839]: I don't understand why `findOrOpen` is necessary. It only applies to `IModelReadRpc`, but I don't think it's right there either. Why is that necessary for reattaching a token?

**Review by kabentley — CHANGES_REQUESTED** [review]

File: core/backend/src/CheckpointManager.ts [hunk 7]
```diff
@@ -112,6 +112,8 @@ export class Downloads {
  * @internal
 */
 export class V2CheckpointManager {
+  private static reattachDueTimestamps: { [key: string]: number } = {};
```
[the remark this answers is not in the document]
Comment (paulius-valiunas) [reply core/backend/src/CheckpointManager.ts:115]: Well, first of all I didn't want to add a new public property that doesn't apply to most checkpoints. If I were to add it there, I'd rather make a new sub-class for V2 checkpoints and use type checking instead of `isV2Checkpoint`. But even then, I don't have a `SnapshotDb` object in `V2CheckpointManager.attach`, I only have the props. To make it worse, when opening a V2 checkpoint, `attach` has to be called before the `SnapshotDb` is constructed, so I wouldn't be able to use `findByKey`.

[the remark this answers is not in the document]
Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:115]: Please remove all of the findOrOpen stuff from IModelDb - that's not appropriate for anything other than RpcRead cases (and even then, i don't understand when it would ever come into play). 

Just add a `reattachDaemon` method on IModelDb that does nothing. In SnapshotDb add the timeout member when it is a V2 checkpoint and override `reattachDaemon` to update the daemon if the timeout has expired, otherwise do nothing. In `IModelReadRpcImp` add a method that finds the iModelDb and then always calls `reattachDaemon`. 

BTW, @swbsi I don't understand the `ClientReqeustContext.current` voodoo in RPC. Where does that get set?

[the remark this answers is not in the document]
Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:115]: actually, i see `SnapshotDb.reattachDaemon` already exists. Just update it to check the timeout and add `reattachDaemon` to IModelDb so you can call it without having to do an instanceof test. 

Reattaching daemon may become relevant for `BriefcaseDb` for web editing backends sometime in the future (although they won't use the CheckpointManger to get the SAS token).

[the remark this answers is not in the document]
Comment (swbsi) [reply core/backend/src/CheckpointManager.ts:115]: > BTW, @swbsi I don't understand the `ClientReqeustContext.current` voodoo in RPC. Where does that get set?

It's set in RpcInvocation. The primary purpose of the request context is to propgate the request's activity id through the awaits that might occur in an RPC implementation function.

https://github.com/imodeljs/imodeljs/blob/2f92222e17eb2cd7ab0e8b09447b06a12de6ade9/core/common/src/rpc/core/RpcInvocation.ts#L123

https://github.com/imodeljs/imodeljs/blob/master/docs/learning/backend/ManagingClientRequestContext.md

[the remark this answers is not in the document]
Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:115]: >  The primary purpose of the request context is to propgate the request's activity id 

in this case we're using it to get the AccessToken for the user from the frontend. I assume this is the right way?

**Review by kabentley — APPROVED** [review]

File: core/backend/src/CheckpointManager.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (wgoehrig) [reply core/backend/src/CheckpointManager.ts:115]: @paulius-valiunas this line (`reattachDueTimestamps`) should still be removed, right?

File: common/scripts/install-run-rush.js [hunk 8]
```diff
@@ -1,3 +1,7 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (paulius-valiunas) [reply common/scripts/install-run-rush.js]: apparently my git commit hook is broken

File: core/backend/src/rpc-impl/RpcBriefcaseUtility.ts [hunk 9]
```diff
@@ -95,6 +95,15 @@ export class RpcBriefcaseUtility {
     }
   }
 
+  public static async findOrOpen(requestContext: AuthorizedClientRequestContext, iModel: IModelRpcProps, syncMode: SyncMode): Promise<IModelDb> {
+    const iModelDb = IModelDb.tryFindByKey(iModel.key);
+    if (undefined === iModelDb) {
+      return this.open({ requestContext, tokenProps: iModel, syncMode, timeout: 1000 });
+    }
+    await iModelDb.reattachDaemon(requestContext);
```
[the remark this answers is not in the document]
Comment (paulius-valiunas) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:103]: well, since the only thing after the await is `return iModel`, this won't change anything, but if we're following the rules strictly, yes, there should be a call to enter here.

[the remark this answers is not in the document]
Comment (wgoehrig) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:103]: Well really my concern is with all the awaits in RPC method implementations where you call this - if we just add the extra enter here, we shouldn't need to do it after every call to `findOrOpen` in `IModelReadRpcImpl`, `TileRequestMemoizer`, etc.

Without changing one of the two (here or each call site), I think the ActivityID can get out-of-sync...

[the remark this answers is not in the document]
Comment (paulius-valiunas) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:103]: Ah, right. I forgot to add it to the RPC implementation classes. But if I understand the event loop correctly, just adding it here wouldn't solve it: the scheduler can queue microtasks after the return statement here but before continuing the method that called this one. That could throw off the activity id too.

[the remark this answers is not in the document]
Comment (kabentley) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:103]: just a comment on the whole `requestContext.enter()` nonsense. Having a global variable that holds the "current context" is a complete disaster and should be removed. It is only there for logging, and our whole strategy for logging is backwards. We should log when we catch errors, *not* when we throw them, and then we should always pass whatever "context" is necessary from the arguments to the failing function. Auto-logging on throws is wrong, and having "magic context" for logging is even worse.

Sorry @paulius-valiunas this has nothing to do with this PR, but all the requestContext.enter crap is annoying as hell to me.

[the remark this answers is not in the document]
Comment (calebmshafer) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:103]: @kabentley I agree on the pain points of `requestContext.enter()` and it's something I'd like to visit for 3.0 as it has implication on decoupling things as well.

File: core/backend/src/CheckpointManager.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulius-valiunas) [reply core/backend/src/CheckpointManager.ts]: should I mark this as resolved? I assume this is the right way to get the access token since that's how other RPC implementation methods do it.

File: core/backend/src/rpc-impl/RpcBriefcaseUtility.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (paulius-valiunas) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:103]: @wgoehrig I added `requestContext.enter()` to every method that uses `findOrOpen`, we should be good now

**Review by wgoehrig — APPROVED** [review]

**Comment by calebmshafer** [comment]
@Mergifyio backport release/2.17.x
